### PR TITLE
[INFINITY-1485] Fix update plan for HDFS

### PR DESCRIPTION
**Pull request by gabrielhartmann**

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/Constants.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
     /** The name used for the task deployment plan. */
     public static final String DEPLOY_PLAN_NAME = "deploy";
+    public static final String UPDATE_PLAN_NAME = "update";
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/Constants.java]: javadoc

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 2]
```diff
@@ -386,8 +387,47 @@ public DefaultScheduler build() {
...
+         */
+        public static Collection<Plan> overrideDeployPlan(
+                Collection<Plan> plans,
+                ConfigurationUpdater.UpdateResult updateResult) {
+
+            Optional<Plan> updatePlanOptional = plans.stream()
+                    .filter(plan -> plan.getName().equals(Constants.UPDATE_PLAN_NAME))
+                    .findFirst();
+
+            LOGGER.info("Update type: " + updateResult.getUpdateType().name());
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: may as well have these on a single line

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java [hunk 3]
```diff
@@ -20,16 +20,35 @@
...
          */
         public final UUID targetId;
+        private final UpdateType updateType;
+
+        /**
+         * Two types of deployments are differentiated by this type.  Either a services is
+         * being deployed for the first time "DEPLOY", or it is updating a previously deployed
+         * version of the service, "UPDATE"
+         */
+        public enum UpdateType {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java]: `DeploymentType`?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java [hunk 4]
```diff
@@ -117,9 +118,14 @@ public UpdateResult updateConfiguration(ServiceSpec candidateConfig) throws Conf
                                 "%d Errors: %s", errors.size(), sj.toString()));
             }
         } else if (targetConfig == null || !configComparator.equals(targetConfig, candidateConfig)) {
-            LOGGER.info("Changes detected between current target configuration '{}' and new " +
-                            "configuration. Setting target to new configuration.",
-                    targetConfigId);
+            if (targetConfig == null) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java]: If the scheduler gets restarted while still in the middle of an initial deployment, will this result in the remainder of that deployment appearing to be an 'update'?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java [hunk 5]
```diff
@@ -20,16 +20,35 @@
...
+                UUID targetId,
+                UpdateType updateType,
+                Collection<ConfigValidationError> errors) {
             this.targetId = targetId;
             this.errors = errors;
+            this.updateType = updateType;
+        }
+
+        public UpdateType getUpdateType() {
+            return  updateType;
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java]: double space

while in here, consider switching to accessors for targetId/errors as well

**Comment by benclarkwood** [comment]
retest this please

**Comment by gabrielhartmann** [comment]
@nickbp:  Addressed comments.  Let me know what you think.  Will add docs in another PR as this is blocking #964

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/Constants.java [hunk 6]
```diff
@@ -5,8 +5,10 @@
  */
 public class Constants {
 
-    /** The name used for the task deployment plan. */
+    /** The name used for the deployment plan. */
     public static final String DEPLOY_PLAN_NAME = "deploy";
+    /** The name used for the update plan. */
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/Constants.java]: Inconsistent white space.

**Comment by gabrielhartmann** [comment]
retest this please

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java [hunk 7]
```diff
@@ -305,4 +310,18 @@ private void clearConfigsNotListed(Set<UUID> neededConfigs) throws ConfigStoreEx
             configStore.clear(configToClear);
         }
     }
+
+    private UpdateResult.DeploymentType getLastCompletedUpdateType() {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java]: Move to StateStoreUtils to be adjacent to e.g. setSuppressed/isSuppressed which is doing something similar

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java [hunk 8]
```diff
@@ -305,4 +310,18 @@ private void clearConfigsNotListed(Set<UUID> neededConfigs) throws ConfigStoreEx
...
+        Optional<String> keyOptional = stateStore.fetchPropertyKeys().stream()
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/DefaultConfigurationUpdater.java]: Any reason to not just fetch the key you want to check? See e.g. StateStoreUtils.isSuppressed()

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java [hunk 9]
```diff
@@ -16,20 +16,51 @@
      * The result of an {@link ConfigurationUpdater#updateConfiguration(Configuration)} call.
      */
     class UpdateResult {
+        public static final String LAST_COMPLETED_UPDATE_TYPE_KEY = "last-completed-update-type";
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java]: May as well put this as a private const in StateStoreUtils as that's now the only thing using it

Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/ConfigurationUpdater.java]: Thanks.  Missed that.

**Pull request merged**